File: src/FileGDB/samplesC#/ExecutingSQL/ExecutingSQL.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Esri.FileGDB;

namespace ExecutingSQL
{
    //
    // Sample: ExecutingSQL
    //
    // Demonstrates how to use ExecuteSQL to perform standard SQL requests on a file
    // geodatabase.

/*
   Copyright © 2025 Esri

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

   A local copy of the license and additional notices are located with the
   source distribution at:

   http://github.com/Esri/file-geodatabase-api/FileGDB_API_1.5.1
*/

  class ExecutingSQL
  {
    static void Main(string[] args)
    {
      try
      {
        // Open the geodatabase.
        Geodatabase geodatabase = Geodatabase.Open("../../samples/data/ExecuteSQL.gdb");

        // Perform a simple SELECT: find the names of every city with a 'TERM' value
        // equal to 'City'.

        string sqlStatement = "SELECT CITY_NAME, POP1990 FROM Cities WHERE TYPE = 'city' AND OBJECTID < 10";
        foreach (Row attrQueryRow in geodatabase.ExecuteSQL(sqlStatement))
        {
          Console.WriteLine("{0}\t{1}", attrQueryRow.GetString("CITY_NAME"), attrQueryRow.GetInteger("POP1990"));
        }

        // SELECT * - Return all fields discover the names and types from the
        // the row.
        sqlStatement = "SELECT * FROM Cities WHERE TYPE = 'city' AND OBJECTID < 10";
        foreach (Row attrQueryRow in geodatabase.ExecuteSQL(sqlStatement))
        {
          for (int nFieldNumber = 0; nFieldNumber < attrQueryRow.FieldInformation.Count; nFieldNumber++)
          {
            string fieldName = attrQueryRow.FieldInformation.GetFieldName(nFieldNumber);
            if (attrQueryRow.IsNull(fieldName))
            {
              Console.Write("null\t");
              continue;
            }

            //Console.Write("{0}", attrQueryRow[fieldName]);
            switch (attrQueryRow.FieldInformation.GetFieldType(nFieldNumber))
            {
              case FieldType.SmallInteger:
                Console.WriteLine("{0}", attrQueryRow.GetShort(fieldName));
                break;
              case FieldType.Integer:
                Console.Write("{0}\t", attrQueryRow.GetInteger(fieldName));
                break;
              case FieldType.Single:
                Console.Write("{0}\t", attrQueryRow.GetFloat(fieldName));
                break;
              case FieldType.Double:
                Console.Write("{0}\t", attrQueryRow.GetDouble(fieldName));
                break;
              case FieldType.String:
                Console.Write("{0}\t", attrQueryRow.GetString(fieldName));
                break;
              case FieldType.Date:
                Console.Write("{0}\t", attrQueryRow.GetDate(fieldName).ToShortTimeString());
                break;
              case FieldType.OID:
                Console.Write("{0}\t", attrQueryRow.GetOID());
                break;
              case FieldType.Geometry:
                Console.Write("Geometry\t");
                break;
              case FieldType.Blob:
                Console.Write("Blob\t");
                break;
              case FieldType.GUID:
                Console.Write("{0}\t", attrQueryRow.GetGUID(fieldName).ToString());
                break;
              case FieldType.GlobalID:
                Console.Write("{0}\t", attrQueryRow.GetGlobalID().ToString());
                break;
              default:
                break;
            }
          }
          Console.WriteLine("");
        }

        // Close the geodatabase
        geodatabase.Close();
      }
      catch (FileGDBException ex)
      {
        Console.WriteLine("{0} - {1}", ex.Message, ex.ErrorCode);
      }
      catch (Exception ex)
      {
        Console.WriteLine("General exception.  " + ex.Message);
      }
    }
  }
}

File: src/FileGDB/samples/ExecutingSQL/ExecutingSQL.cpp
//
// Sample: ExecutingSQL
//
// Demonstrates how to use ExecuteSQL to perform standard SQL requests on a file
// geodatabase.  Prior to running this example, copy the ExecuteSQL.gdb from the
// data directory to the working directory.

// Copyright 2015 ESRI
// 
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
// 
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
// 
// See the use restrictions at <your File Geodatabase API install location>/license/userestrictions.txt.
// 

#include <string>
#include <iostream>
#include <fstream>
#include <time.h>

#include <FileGDBAPI.h>

using namespace std;
using namespace FileGDBAPI;

int main()
{
  // Open the geodatabase.
  fgdbError   hr;
  wstring     errorText;
  Geodatabase geodatabase;
  if ((hr = OpenGeodatabase(L"../data/ExecuteSQL.gdb", geodatabase)) != S_OK)
  {
    wcout << "An error occurred while opening the geodatabase." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  // Perform a simple SELECT: find the names of every city with a 'TYPE' value
  // equal to 'city'.
  wstring sqlStatement(L"SELECT CITY_NAME, POP1990 FROM Cities WHERE TYPE = 'city' AND OBJECTID < 10");
  EnumRows attrQueryRows;
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the attribute query." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  // Iterate through the returned rows.
  Row     attrQueryRow;
  int32   cityPop;
  wstring cityName;
  while (attrQueryRows.Next(attrQueryRow) == S_OK)
  {
    attrQueryRow.GetInteger(L"POP1990", cityPop);
    attrQueryRow.GetString(L"CITY_NAME", cityName);

    wcout << cityName << '\t' << cityPop << endl;
  }

  // SELECT * - Return all fields.
  sqlStatement.assign(L"SELECT * FROM Cities WHERE TYPE = 'city' AND OBJECTID < 10");
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the attribute query." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  // Get the field type and name from the row enumerator.
  FieldInfo fieldInfo;
  attrQueryRows.GetFieldInformation(fieldInfo);

  int       fieldCount;
  FieldType fieldType;
  wstring   fieldName;

  // Iterate through the returned rows printing out all field values.
  short     shortField;
  int32     longField;
  float     floatField;
  double    doubleField;
  string    stringField;
  wstring   wstringField;
  tm        dateTimeField;
  char      datetime[80];
  Guid      globalIDField;
  Guid      guidField;
  wstring   strGuid;
  wstring   strGlobalID;
  bool      isNull;

  while (attrQueryRows.Next(attrQueryRow) == S_OK)
  {
    fieldInfo.GetFieldCount(fieldCount);
    for (long fieldNumber = 0; fieldNumber < fieldCount; fieldNumber++)
    {
      fieldInfo.GetFieldType(fieldNumber, fieldType);
      fieldInfo.GetFieldName(fieldNumber, fieldName);
      attrQueryRow.IsNull(fieldName, isNull);
      if (!isNull)
      {
        switch (fieldType)
        {
          case fieldTypeSmallInteger:
            attrQueryRow.GetShort(fieldName, shortField);
            wcout <<  shortField << endl;
          break;

          case fieldTypeInteger:
            attrQueryRow.GetInteger(fieldName, longField);
            wcout << longField << '\t';
          break;

          case fieldTypeSingle:
            attrQueryRow.GetFloat(fieldName, floatField);
            wcout << floatField << '\t';
          break;

          case fieldTypeDouble:
            attrQueryRow.GetDouble(fieldName, doubleField);
            wcout << doubleField << '\t';
          break;

          case fieldTypeString:
            attrQueryRow.GetString(fieldName, wstringField);
            wcout << wstringField << '\t';
          break;

          case fieldTypeDate:
            attrQueryRow.GetDate(fieldName, dateTimeField);
            strftime(datetime,80,"%a %b %d %I:%M:%S%p %Y", &dateTimeField);
            wcout << datetime << '\t';
          break;

          case fieldTypeOID:
            attrQueryRow.GetOID(longField);
            wcout << longField << '\t';
          break;

          case fieldTypeGeometry:
            wcout << "Geometry" << '\t';
          break;

          case fieldTypeBlob:
            wcout << "Blob" << '\t';
          break;

          case fieldTypeGUID:
            attrQueryRow.GetGUID(fieldName, guidField);
            guidField.ToString(strGuid);
            wcout << strGuid << '\t';
          break;

          case fieldTypeGlobalID:
            attrQueryRow.GetGlobalID(globalIDField);
            globalIDField.ToString(strGlobalID);
            wcout << strGlobalID << '\t';
          break;

          default:
          break;
        }
      }
      else
      {
        wcout << "null" << '\t';
      }
    }

    wcout << endl;
  }

  // Perform an UPDATE
  sqlStatement.assign(L"UPDATE CitiesNoGeo SET POP1990 = 65678 WHERE STATE_CITY = '0659962'");
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the update." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  sqlStatement.assign(L"SELECT CITY_NAME, POP1990 FROM CitiesNoGeo WHERE STATE_CITY = '0659962'");
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the attribute query." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  // Iterate through the returned rows.
  while (attrQueryRows.Next(attrQueryRow) == S_OK)
  {
    attrQueryRow.GetInteger(L"POP1990", cityPop);
    attrQueryRow.GetString(L"CITY_NAME", cityName);

    wcout << "Changed the population of " << cityName << " from 60394 to "  << cityPop << endl;
  }

  // Perform another UPDATE
  sqlStatement.assign(L"UPDATE CitiesNoGeo SET POP1990 = 60394 WHERE STATE_CITY = '0659962'");
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the update." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  sqlStatement.assign(L"SELECT CITY_NAME, POP1990 FROM CitiesNoGeo WHERE STATE_CITY = '0659962'");
  if ((hr = geodatabase.ExecuteSQL(sqlStatement, true, attrQueryRows)) != S_OK)
  {
    wcout << "An error occurred while performing the attribute query." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  // Iterate through the returned rows.
  while (attrQueryRows.Next(attrQueryRow) == S_OK)
  {
    attrQueryRow.GetInteger(L"POP1990", cityPop);
    attrQueryRow.GetString(L"CITY_NAME", cityName);

    wcout << "Changed the population of " << cityName << " back to "  << cityPop << endl;
  }

  attrQueryRows.Close(); // Close the EnumRows

  // Close the geodatabase
  if ((hr = CloseGeodatabase(geodatabase)) != S_OK)
  {
    wcout << "An error occurred while closing the geodatabase." << endl;
    ErrorInfo::GetErrorDescription(hr, errorText);
    wcout << errorText << "(" << hr << ")." << endl;
    return -1;
  }

  return 0;
}
